src/lib/math/numbertheory/mod_inv.cpp

```diff
@@ -277,4 +279,100 @@ BigInt inverse_mod(const BigInt& n, const BigInt& mod) {
    return h;
 }
 
+BigInt inverse_modulo_secret_prime(const BigInt& x, const BigInt& p) {
+   BOTAN_ARG_CHECK(x.is_positive() && p.is_positive(), "Parameters must be positive");
+   BOTAN_ARG_CHECK(x < p, "x must be less than p");
+   BOTAN_ARG_CHECK(p.is_odd() and p > 1, "Primes are odd integers greater than 1");
+
+   // TODO possibly use FLT, or the algorithm presented for this case in
+   // Handbook of Elliptic and Hyperelliptic Curve Cryptography
+
+   return inverse_mod_odd_modulus(x, p);
+}
+
+BigInt inverse_modulo_public_prime(const BigInt& x, const BigInt& p) {
+   return inverse_modulo_secret_prime(x, p);
+}
+
+BigInt inverse_modulo_rsa_public_modulus(const BigInt& x, const BigInt& n) {
+   BOTAN_ARG_CHECK(n.is_positive() && n.is_odd(), "RSA public modulus must be odd and positive");
+   BOTAN_ARG_CHECK(x.is_positive() and x < n, "Input must be positive and less than RSA modulus");
+   BigInt z = inverse_mod_odd_modulus(x, n);
+   BOTAN_ASSERT(!z.is_zero(), "Accidentally factored the public modulus");  // whoops
+   return z;
+}
+
+BigInt compute_rsa_secret_exponent(const BigInt& e, const BigInt& phi_n, const BigInt& p, const BigInt& q) {
+   /*
+   * Both p - 1 and q - 1 are chosen to be relatively prime to e. Thus
+   * phi(n), the least common multiple of p - 1 and q - 1, is also
+   * relatively prime to e.
+   */
+   BOTAN_DEBUG_ASSERT(gcd(e, phi_n) == 1);
+
+   if(e == 65537) {
+      /*
+      Arazi's algorithm for inversion of prime x modulo a non-prime
+
+      "GCD-Free Algorithms for Computing Modular Inverses"
+      Marc Joye and Pascal Paillier, CHES 2003 (LNCS 2779)
+      https://marcjoye.github.io/papers/JP03gcdfree.pdf
+
+      This could be extended to cover other cases such as e=3 or e=17 but
+      these days 65537 is the standard RSA public exponent
+      */
+
+      const word phi_mod_e = ct_mod_word(phi_n, 65537);
+
+      // Compute phi^-1 mod e using FLT
+
+      const word inv_phi_mod_e = [&]() {
+         // Need 64-bit here as accum*accum exceeds 32-bit if accum=0x10000
+         uint64_t accum = 1;
+         for(size_t i = 0; i != 16; ++i) {
```

Comment:
TODO here - these reductions on x86 and ARM compile down to just multiplication by a constant, shifts/adds, etc. But on RISCV it branches to some runtime function which is likely not constant time.

Reply:
Fixed to use a word-length Barrett reduction